src/Common/ProxyListConfigurationResolver.cpp

```diff
@@ -0,0 +1,31 @@
+#include <Common/ProxyListConfigurationResolver.h>
+
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/logger_useful.h>
+#include <Poco/URI.h>
+
+namespace DB
+{
+
+ProxyListConfigurationResolver::ProxyListConfigurationResolver(std::vector<Poco::URI> proxies_)
+: proxies(std::move(proxies_))
+{
+}
+
+ProxyConfiguration ProxyListConfigurationResolver::resolve(bool)
```

Comment:
IMO it shouldn't just give away just any proxy from the list if caller requested HTTPS proxy. If there is no HTTPS proxy it is OK to throw an exception.

Safer way to design this interface would be to have an `enum {HTTP, HTTPS, ANY}` instead of just `bool` and make default value to `ANY`.

So that when caller doesn't care about type of proxy returned, it calls `resolver.resolve(ANY)` or just `resolver.resolve()`.

Reply:
I agree with you, I asked myself the same thing when I saw this code (remember, this already existed and I simply adapted / moved it).

This is probably a bug in the original implementation OR the developer simply relied on the idea that usually http and https proxies are the same.

I'll see what I can do about it.

Reply:
Couple of thoughts about this:

Current List and Remote resolvers do not care if the request is HTTP or HTTPs, it'll return any. By definition, it is wrong, but probably works for most cases since it's usually the same server.

1. **The list settings syntax does not allow the user to specify the HTTP method, so we can't solve this problem without modifying it, there is no way to infer it in the code**. _To fix this, we can introduce something like: <uri_s></uri_s> that allows the user to specify if it's a HTTP or HTTPs proxy. We'll be assuming the existing / old settings are all about HTTP proxies and pass Method::ANY in that specific call to make it backwards compatible_.
2. **The remote resolver allows the user to specify the method, but limits it to only one resolver, so the user can't have both**. _Removing that limit will do the trick, but will also allow the user to specify two or even more HTTP only resolvers. I guess that's totally ok, if the user does not specify HTTPs, we simply assume there is no https. What do we do with the multiple HTTP remote resolvers? Probably just round-robin it like the list resolver does (increases the impl complexity a bit..)_.

**Note**: It is a bit tricky to round-robin across 3 different method options: HTTP, HTTPs and ANY. Lots of state will have to be kept and implementation will become a bit more complex.